pyke/lightcurve.py

```diff
@@ -150,6 +179,101 @@ def detrend(self):
         """
         pass
 
+class SystematicsCorrector(object):
+    def correct(self):
+        pass
+
+
+class KeplerCBVCorrector(SystematicsCorrector):
+    """Remove systematic trends Kepler light curves by fitting
+    cotrending basis vectors.
+
+    .. math::
+
+         \arg \min_{\theta \in \Theta} |f(t) - <\theta, \left[\rm{cbv}_1(t), ..., \rm{cbv}_n(t)\right]^{T}|^p
+
+    Attributes
+    ----------
+    lc_file : KeplerLightCurveFile object or str
+    cbvs : list of ints
+
+    Notes
+    -----
+    The cotrending basis vectors files can be found
+    here: http://archive.stsci.edu/missions/kepler/cbv/
+    """
+
+    def __init__(self, lc_file, cbvs=[1, 2]):
+        self.lc_file = lc_file
+        self.cbvs = cbvs
+
+        if self.lc_file.mission == 'Kepler':
+            self.cbv_base_url = "http://archive.stsci.edu/missions/kepler/cbv/"
+        elif self.lc_file.mission == 'K2':
+            self.cbv_base_url = "http://archive.stsci.edu/missions/k2/cbv/"
+
+    @property
+    def lc_file(self):
+        return self._lc_file
+
+    @lc_file.setter
+    def lc_file(self, value):
+        if isinstance(value, str):
+            self._lc_file = KeplerLightCurveFile(value)
+        elif isinstance(value, KeplerLightCurveFile):
+            self._lc_file = value
+
+    def correct(self):
+        module, output = channel_to_module_output(self.lc_file.channel)
+        cbv_file = pyfits.open(self.get_cbv_file())
+        cbv_data = cbv_file['MODOUT_{0}_{1}'.format(module, output)].data
+
+        cbv_array = []
+        for i in self.cbvs:
+            cbv_array.append(cbv_data.field('VECTOR_{}'.format(i)))
+        cbv_array = np.asarray(cbv_array)
+
+        sap_lc = self.lc_file.SAP_FLUX
+        median_sap_flux = np.nanmedian(sap_lc.flux)
+        norm_sap_flux = sap_lc.flux / median_sap_flux - 1
+        norm_err_sap_flux = (sap_lc.flux_err / median_sap_flux) ** 2
+
+        def mean_model(*theta):
+            coeffs = np.asarray(theta)
+            return np.dot(coeffs, cbv_array)
+
+        chi_sqr = oktopus.GaussianLikelihood(data=norm_sap_flux,
+                                             mean=mean_model,
```

Comment:
Do you think we should use a Laplacian distribution here, i.e. which yields the L1 norm (obviously)? 
Do add a comment to explain that it does the L1 norm thing, because some people may not know.

Reply:
🍓 🍓 🍓 🍓